### ISSUE TITLE
Add support to JWT Token

### ISSUE THREAD

**Issue by lucasdss** [DarthSim/imgproxy#107]
A JWT token signature validation would allow imgproxy to be added in the microservices architecture smoothly.

**Comment by DarthSim** [comment]
Could you explain the benefits of JWT over the current solution? I kept JWT in mind from the start, but I still don't think this is what imgproxy needs.

**Comment by lucasdss** [comment]
I was thinking to add imgproxy as an internal component behind a session server, for example. With a Nginx in the middle, it's easy to do something like that, but I would prefer to do not use another component. So to validate that every request comes from a trustable component I think would be nice to validate the token signature. So you don't need to have a fixed token but just a public key, assuming it's an RS256.

For sure it's not a must, but a nice to have.

**Comment by DarthSim** [comment]
imgproxy uses [URL signature](https://github.com/DarthSim/imgproxy/blob/master/docs/signing_the_url.md) for validating incoming requests. Practically, the same thing as used in JWT but without the overhead like header and encoding data with Base64.

**Comment by knpwrs** [comment]
I just discovered imgproxy and was wondering about JWT support myself. A major advantage of JWTs is that they are standard. Rather than having to implement URL signing for various different services (imgproxy, s3/aws, various CDNs, etc) one can just stick all of their claims into a JWT and have everything validated in a standard way. JWTs would also have built-in expiration -- I can see that imgproxy has its own scheme for expiration but having features like this standardized across different services is a major selling point in my humble opinion.

**Comment by DarthSim** [comment]
Hey @knpwrs!

There are some reasons why I don't like the idea of JWT support:

1. You can't figure out what processing options are used by looking at JWT token. Processing options are not a secret thing, so there's no sense in hiding them, and it's handy for a developer to see them right in the URL. We added plain URL support for the same reason.
2. We tend to keep imgproxy as simple as possible. We dropped basic URL format support from imgproxy v3 as it was pretty confusing to have multiple URL formats. Honestly, I don't want to bring this confusion back. Also, JWT will hardly be popular because of the first reason. In a matter of fact, you are only the second person who requested this in 5 years.
3. It's easy to shoot yourself in the leg with JWT. For example, using RSA signing method will give you different signatures every time you sign the same data. Because of this, requests to imgproxy will always miss the CDN's cache. We could limit signing methods to HMAC only, but this would drop half of JWT pros.
4. URL signing process in imgproxy is pretty simple, and JWT looks like a complication here.
5. Speaking of standards. While JWT is definitely the standard for data exchange, in the image servers world almost everyone signs URLs pretty the same way as imgproxy.

**Comment by jeaneric** [comment]
Hi,

I personnaly think that the JWT option would be good. But not for signing the request, there are too many disadvantages as @DarthSim point out.

I would see the JWT just to ensure that an "insecure" request is submitted by a user authenticated by a JWT. Here what I see:

- The JWT does not contain any of the resquest, it just ensure that the user is authenticated. 
- JWT could be sent as many app support it in the Bearer Authentication header. For my use cas, it would be more useful if it would be also supported as a query parameter in the URL (?jwt-token=.....).
- At first the JWT could only be use to ensure a user have rights to submit a request, but it could aslo contains information about operation that the user is allowed to perform, as one user might be able to resquest crop image larger than 600px, and not the other one. It remove the hurdle of managing user and rights by imgProxy, it just ensure that the JWT token and it's informations allow the resquest.
- The "insecure" part of the request could also be replaced by a "jwt-auth" word that would trigger the JWT validation.

**Comment by DarthSim** [comment]
Hey @jeaneric!

This looks very much like something that should be implemented outside of imgproxy.

**Comment by jeaneric** [comment]
> Hey @jeaneric!
> 
> This looks very much like something that should be implemented outside of imgproxy.

You are totally right, it could be integrated inside a proxy for example. As @lucasdss said in the [third comment in the thread](https://github.com/imgproxy/imgproxy/issues/107#issuecomment-442013179), it's not fundamental, but it simplify the stack for those already using JWT. In it's most basic form, it's just a signature validation, but I'm not a Go programmer.
